File: src/main/kotlin/com/iamalangreen/self/lifting/workout.kt
package com.iamalangreen.self.lifting

import com.fasterxml.jackson.annotation.JsonFormat
import com.iamalangreen.self.Response
import com.iamalangreen.self.success
import jakarta.persistence.*
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId

data class WorkoutRequest(
    val id: Long?,
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
    val startTime: LocalDateTime?,
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
    val endTime: LocalDateTime?,
    val gym: Long,
    val routine: Long?,
    val target: Set<Long> = setOf(),
    val note: String?
)

data class WorkoutResponse(
    val id: Long,
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
    val startTime: LocalDateTime?,
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm")
    val endTime: LocalDateTime?,
    val gym: GymResponse,
    val routine: RoutineResponse?,
    val target: Set<TargetResponse> = setOf(),
    val note: String?
)

fun Workout.toResponse(): WorkoutResponse {
    return WorkoutResponse(
        id!!,
        startTime,
        endTime,
        gym.toResponse(),
        routine?.toRoutineResponse(),
        target.map { it.toResponse() }.toSet(),
        note
    )
}

@RestController
@RequestMapping("/api/lifting/workout")
class WorkoutController(val workoutService: WorkoutService, val gymService: GymService) {

    @PostMapping
    fun createWorkout(request: WorkoutRequest): Response {
        val workout =
            workoutService.create(request.startTime, request.gym, request.routine, request.target, request.note)
        return success(workout.toResponse())
    }

    @GetMapping
    fun getAllWorkout(): Response {
        return success(workoutService.getAll().map { it.toResponse() })
    }

    @PostMapping("/stop")
    fun stopWorkout(request: WorkoutRequest): Response {
        val workout = workoutService.update(
            request.id!!,
            request.startTime,
            request.endTime,
            request.gym,
            request.routine,
            request.target,
            request.note
        )
        return success(workout.toResponse())
    }

    @GetMapping("/in-process")
    fun findInProcessWorkout(): Response {
        return success(workoutService.findInProcessWorkout().toResponse())
    }

}

@Service
class WorkoutService(
    private val workoutRepository: WorkoutRepository,
    private val targetService: TargetService,
    private val gymService: GymService,
    private val routineService: RoutineService
) {
    fun create(
        startTime: LocalDateTime?,
        gymId: Long,
        routineId: Long?,
        targetIds: Set<Long>,
        note: String?
    ): Workout {
        val gym = gymService.getById(gymId)
        val routine = routineId?.let { routineService.getById(routineId) }
        val targets = targetIds.map { targetService.getById(it) }.toMutableSet()
        val workout = workoutRepository.save(
            Workout(
                startTime = startTime,
                endTime = null,
                gym = gym,
                routine = routine,
                target = targets,
                note = note
            )
        )
        return workout
    }

    fun getAll(): List<Workout> {
        return workoutRepository.findAll()
    }

    fun update(
        id: Long,
        startTime: LocalDateTime?,
        endTime: LocalDateTime?,
        gym: Long,
        routine: Long?,
        target: Set<Long>,
        note: String?
    ): Workout {
        val workout = workoutRepository.findById(id).orElseThrow()
        if (startTime != workout.startTime)
            workout.startTime = startTime
        if (endTime != workout.endTime)
            workout.endTime = endTime
        if (gym != workout.gym.id)
            workout.gym = gymService.getById(gym)
        if (routine != null && routine != workout.routine?.id)
            workout.routine = routineService.getById(routine)
        if (target.isNotEmpty()) {
            workout.target.clear()
            workout.target.addAll(target.map { targetService.getById(it) })
        }
        return workoutRepository.save(workout)
    }

    fun findInProcessWorkout(): Workout {
        val workouts = getAll()
        val today = LocalDate.now(ZoneId.of("Asia/Shanghai"))
        val workout = workouts.first { it.startTime?.toLocalDate()?.equals(today) ?: false }
        return workout
    }
}

interface WorkoutRepository : JpaRepository<Workout, Long> {}

@Entity
@Table(name = "lifting_workout")
data class Workout(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
    @Column
    var startTime: LocalDateTime?,
    @Column
    var endTime: LocalDateTime?,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "gym_id")
    var gym: Gym,
    @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinColumn(name = "routine_id")
    var routine: Routine?,
    @ManyToMany(cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    @JoinTable(
        name = "lifting_target",
        joinColumns = [JoinColumn(name = "workout_id")],
        inverseJoinColumns = [JoinColumn(name = "target_id")]
    )
    var target: MutableSet<Target> = mutableSetOf(),
    @Column
    var note: String?
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Workout

        return id == other.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }

}

File: backend/app/api/workout.py
from typing import Optional
from datetime import datetime
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.services.workout import WorkoutService
from app.schemas.workout import WorkoutRequest, WorkoutResponse, WorkoutSummaryResponse
from app.schemas.gym import GymResponse
from app.schemas.routine import RoutineSummaryResponse
from app.schemas.target import TargetResponse
from app.schemas.response import Response, PageResponse

router = APIRouter()


@router.post("", response_model=Response)
async def create_workout(request: WorkoutRequest, db: Session = Depends(get_db)):
    """创建训练"""
    workout = WorkoutService.create(
        db,
        start_time=request.start_time,
        gym_id=request.gym,
        routine_id=request.routine,
        target_ids=request.target,
        note=request.note
    )
    
    # 构建响应
    gym_response = GymResponse(
        id=workout.gym.id,
        name=workout.gym.name,
        location=workout.gym.location
    )
    
    targets = [TargetResponse.model_validate(target) for target in workout.targets]
    
    response_data = WorkoutResponse(
        id=workout.id,
        start_time=workout.start_time,
        end_time=workout.end_time,
        gym=gym_response,
        routine=None,
        target=set(targets),
        note=workout.note
    )
    
    return Response(data=response_data)


@router.put("/{workout_id}", response_model=Response)
async def update_workout(workout_id: int, request: WorkoutRequest, db: Session = Depends(get_db)):
    """更新训练"""
    workout = WorkoutService.update(
        db,
        workout_id=workout_id,
        start_time=request.start_time,
        end_time=request.end_time,
        gym_id=request.gym,
        routine_id=request.routine,
        target_ids=request.target,
        note=request.note
    )
    
    # 构建响应
    gym_response = GymResponse(
        id=workout.gym.id,
        name=workout.gym.name,
        location=workout.gym.location
    )
    
    # 处理训练计划关联
    routine_response = None
    if workout.routine:
        routine_response = RoutineSummaryResponse(
            id=workout.routine.id,
            name=workout.routine.name,
            description=workout.routine.description,
            targets=set()
        )
    
    targets = [TargetResponse.model_validate(target) for target in workout.targets]
    
    response_data = WorkoutResponse(
        id=workout.id,
        start_time=workout.start_time,
        end_time=workout.end_time,
        gym=gym_response,
        routine=routine_response,
        target=set(targets),
        note=workout.note
    )
    
    return Response(data=response_data)


@router.post("/stop", response_model=Response)
async def stop_workout(request: WorkoutRequest, db: Session = Depends(get_db)):
    """停止训练"""
    workout = WorkoutService.update(
        db,
        workout_id=request.id,
        start_time=request.start_time,
        end_time=request.end_time,
        gym_id=request.gym,
        routine_id=request.routine,
        target_ids=request.target,
        note=request.note
    )
    
    # 构建响应
    gym_response = GymResponse(
        id=workout.gym.id,
        name=workout.gym.name,
        location=workout.gym.location
    )
    
    # 处理训练计划关联
    routine_response = None
    if workout.routine:
        routine_response = RoutineSummaryResponse(
            id=workout.routine.id,
            name=workout.routine.name,
            description=workout.routine.description,
            targets=set()
        )
    
    targets = [TargetResponse.model_validate(target) for target in workout.targets]
    
    response_data = WorkoutResponse(
        id=workout.id,
        start_time=workout.start_time,
        end_time=workout.end_time,
        gym=gym_response,
        routine=routine_response,
        target=set(targets),
        note=workout.note
    )
    
    return Response(data=response_data)


@router.get("", response_model=Response)
async def get_all_workouts(
    page: int = Query(0, ge=0, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    start_date: Optional[datetime] = Query(None, description="开始日期"),
    end_date: Optional[datetime] = Query(None, description="结束日期"),
    db: Session = Depends(get_db)
):
    """获取所有训练"""
    skip = page * size
    workouts = WorkoutService.get_all(db, start_date=start_date, end_date=end_date, skip=skip, limit=size)
    total = WorkoutService.get_total_count(db, start_date=start_date, end_date=end_date)
    
    items = []
    for workout in workouts:
        gym_response = GymResponse(
            id=workout.gym.id,
            name=workout.gym.name,
            location=workout.gym.location
        )
        
        targets = [TargetResponse.model_validate(target) for target in workout.targets]
        
        item = WorkoutResponse(
            id=workout.id,
            start_time=workout.start_time,
            end_time=workout.end_time,
            gym=gym_response,
            routine=None,
            target=set(targets),
            note=workout.note
        )
        items.append(item)
    
    page_response = PageResponse(
        total=total,
        page=page,
        size=size,
        items=items
    )
    
    return Response(data=page_response)


@router.get("/in-process", response_model=Response)
async def find_in_process_workout(db: Session = Depends(get_db)):
    """查找进行中的训练"""
    workout = WorkoutService.find_in_process_workout(db)
    
    if not workout:
        return Response(data=None)
    
    # 构建响应
    gym_response = GymResponse(
        id=workout.gym.id,
        name=workout.gym.name,
        location=workout.gym.location
    )
    
    # 处理训练计划关联
    routine_response = None
    if workout.routine:
        routine_response = RoutineSummaryResponse(
            id=workout.routine.id,
            name=workout.routine.name,
            description=workout.routine.description,
            targets=set()
        )
    
    targets = [TargetResponse.model_validate(target) for target in workout.targets]
    
    response_data = WorkoutResponse(
        id=workout.id,
        start_time=workout.start_time,
        end_time=workout.end_time,
        gym=gym_response,
        routine=routine_response,
        target=set(targets),
        note=workout.note
    )
    
    return Response(data=response_data)
